runtime/src/accounts_db.rs

```diff
@@ -190,6 +190,64 @@ impl CalculateHashIntermediate {
     }
 }
 
+#[derive(Default, Debug)]
+struct CumulativeOffset1D {
+    pub index: usize,
+    pub start_offset: usize,
+    pub len: usize,
+}
+
+impl CumulativeOffset1D {
+    pub fn new(index: usize, start_offset: usize, len: usize) -> CumulativeOffset1D {
+        Self {
+            index,
+            start_offset,
+            len,
+        }
+    }
+}
+
+// Allow retreiving &[start..end] from a logical src: Vec<T>, where src is really Vec<Vec<T>> (or later Vec<Vec<Vec<T>>>)
+// This model prevents callers from having to flatten which saves both working memory and time.
+#[derive(Default, Debug)]
+struct CumulativeOffsets1D {
+    cumulative_offsets: Vec<CumulativeOffset1D>,
+    total_count: usize,
+}
+
+impl CumulativeOffsets1D {
+    pub fn from_raw<T>(raw: &[Vec<T>]) -> CumulativeOffsets1D {
+        let mut total_count: usize = 0;
+        let mut cumulative_offsets = Vec::new();
+        raw.iter().enumerate().for_each(|(i, v)| {
+            let len = v.len();
+            cumulative_offsets.push(CumulativeOffset1D::new(i, total_count, total_count + len));
+            total_count += len;
+        });
```

Comment:
seems like a natural collect
`let cumulative_offsets = raw.iter().enumerate().map(|(i, v)| ...).collect();`

Reply:
yes, and it would be good to remove zero length vectors here, so did that, too